memcopy.h

```diff
@@ -0,0 +1,181 @@
+           inline version that does not incur a function call overhead. */
+        __builtin_memset(out, c, sz);
+        out += sz;
+    }
+
+    return out;
+}
+
+/* Memset LEN bytes in OUT with the value at OUT - 1. Return OUT + LEN. */
+static inline unsigned char *byte_memset_16(unsigned char *out, unsigned len) {
+    unsigned sz = sizeof(long double);
+    unsigned char *from = out - 1;
+    unsigned char c = *from;
+
+    /* assert(len >= sz); */
+
+    /* First, deal with the case when LEN is not a multiple of SZ. */
+    __builtin_memset(out, c, sz);
+    unsigned rem = len % sz;
+    len /= sz;
+    out += rem;
+    from += rem;
+
+    while (len--) {
+        /* When sz is a constant, the compiler replaces __builtin_memset with an
+           inline version that does not incur a function call overhead. */
+        __builtin_memset(out, c, sz);
+        out += sz;
+    }
+
+    return out;
+}
+
+/* Byte by byte semantics: copy LEN bytes from OUT + DIST and write them to OUT. Return OUT + LEN. */
+static inline unsigned char *chunk_memcpy_8(unsigned char *out, unsigned char *from, unsigned len) {
+    long chunk;
+    unsigned sz = sizeof(chunk);
+
+    /* assert(len >= sz); */
+
+    /* Copy a few bytes to make sure the loop below has a multiple of SZ bytes to be copied. */
+    __builtin_memcpy(&chunk, from, sz);
+    __builtin_memcpy(out, &chunk, sz);
+
+    unsigned rem = len % sz;
+    len /= sz;
+    out += rem;
+    from += rem;
+
+    while (len--) {
+        __builtin_memcpy(&chunk, from, sz);
+        __builtin_memcpy(out, &chunk, sz);
+        from += sz;
+        out += sz;
+    }
+
+    return out;
+}
+
+/* Byte by byte semantics: copy LEN bytes from OUT + DIST and write them to OUT. Return OUT + LEN. */
+static inline unsigned char *chunk_memcpy_16(unsigned char *out, unsigned char *from, unsigned len) {
+    long double chunk;
+    unsigned sz = sizeof(chunk);
+
+    /* assert(len >= sz); */
+
+    /* Copy a few bytes to make sure the loop below has a multiple of SZ bytes to be copied. */
+    __builtin_memcpy(&chunk, from, sz);
+    __builtin_memcpy(out, &chunk, sz);
+
+    unsigned rem = len % sz;
+    len /= sz;
+    out += rem;
+    from += rem;
+
+    while (len--) {
+        __builtin_memcpy(&chunk, from, sz);
+        __builtin_memcpy(out, &chunk, sz);
+        from += sz;
+        out += sz;
+    }
+
+    return out;
+}
+
+/* Copy DIST bytes from OUT - DIST into OUT + DIST * k, for 0 <= k < LEN/DIST. Return OUT + LEN. */
+static inline unsigned char *chunk_memset_8(unsigned char *out, unsigned dist, unsigned len) {
+    long chunk = 0;
+    unsigned sz = sizeof(chunk);
+    unsigned char *from = out - dist;
+
+    /* assert(len >= sz); */
+
+    /* Double up the size of the memset pattern until reaching the largest pattern of size less than SZ. */
+    while (dist < len && dist < sz) {
+        __builtin_memcpy(&chunk, from, sz);
+        __builtin_memcpy(out, &chunk, sz);
+
```

Comment:
This is the pattern that appears in Simon's patch, and I think I know a better way to improve its performance:
we should only read DIST bytes and only write SZ bytes.
The memset pattern should be computed in registers with | and <<.
This will improve performance as we will reduce the number of memory accesses.

Let's not commit this patch yet.